### PR TITLE
#307 - Feedback send dialog

### PR DESCRIPTION
**Pull request by nnice**

### CODE REVIEW

**Comment by nnice** [comment]
resolves #307

File: src/test/java/de/qabel/desktop/ui/feedback/FeedbackControllerTest.java [hunk 1]
```diff
@@ -17,16 +22,29 @@ public void createEMailBodyTest() {
         controller.feedbackField.setText("feedback");
 
         controller.handleSendButtonAction();
-        waitUntil(() -> controller.nameField.getText().equals(""));
-        waitUntil(() -> controller.emailField.getText().equals(""));
-        waitUntil(() -> controller.feedbackField.getText().equals(""));
+        runLaterAndWait(() -> {
```
Comment (julianseeger) [inline src/test/java/de/qabel/desktop/ui/feedback/FeedbackControllerTest.java]: does this really need to be inside the ui thread?

File: src/test/java/de/qabel/desktop/ui/feedback/FeedbackControllerTest.java [hunk 2]
```diff
@@ -17,16 +22,29 @@ public void createEMailBodyTest() {
...
+            controller.nameField.getText().equals("");
+            controller.emailField.getText().equals("");
+            controller.feedbackField.getText().equals("");
+        });
     }
 
+    @Test
+    public void createAltertInformations() {
```
Comment (julianseeger) [inline src/test/java/de/qabel/desktop/ui/feedback/FeedbackControllerTest.java]: typo?

File: src/test/java/de/qabel/desktop/ui/feedback/FeedbackControllerTest.java [hunk 3]
```diff
@@ -17,16 +22,29 @@ public void createEMailBodyTest() {
...
+        createController();
+
+        assertEquals(titleBar, controller.titleBar);
+        assertEquals(infoMessage, controller.infoMessage);
```
Comment (julianseeger) [inline src/test/java/de/qabel/desktop/ui/feedback/FeedbackControllerTest.java]: the test does:
```JAVA
controller.titleBar = titleBar;
assertEquals(titleBar, controller.titleBar);
```

where's the point?

File: src/test-gui/java/de/qabel/desktop/ui/feedback/FeedbackGuiTest.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    @Override
+    protected FXMLView getView() {
+        return new FeedbackView();
+    }
+
+    @Test
+    public void sendFeedback() {
+        page.sendFeedback();
```
Comment (julianseeger) [inline src/test-gui/java/de/qabel/desktop/ui/feedback/FeedbackGuiTest.java]: no assertion. This test doesn't test anything.

File: src/main/resources/ui_te_ST.properties [hunk 5]
```diff
@@ -85,3 +85,5 @@ password = Password
 cancel = cancel
 contactImport=Import Contacts
 contactExport=Export Contacts
+feedBackInfoMessage=Thank you for your feedback
+feedBackInfoHeader=Send feedback was successful
```
Comment (julianseeger) [inline src/main/resources/ui_te_ST.properties]: use these properties in the tests. that what they're for. If not, remove this

File: src/main/java/de/qabel/desktop/ui/feedback/FeedbackController.java [hunk 6]
```diff
@@ -47,19 +59,27 @@ public void run() {
                 try {
                     reportHandler.sendFeedback(feedbackField.getText(), nameField.getText(), emailField.getText());
 
-                    feedbackField.setText("");
-                    nameField.setText("");
-                    emailField.setText("");
-                } catch (IOException  e) {
+                    Platform.runLater(() -> {
+                        createInfoBox(infoMessage, titleBar);
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/feedback/FeedbackController.java]: `showThanksDialog` is more descriptive

File: src/main/java/de/qabel/desktop/ui/feedback/FeedbackController.java [hunk 7]
```diff
@@ -47,19 +59,27 @@ public void run() {
...
-        }.start();
-
-
-
 
+        }.start();
     }
-}
 
+    void createInfoBox(String infoMessage, String titleBar) {
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/feedback/FeedbackController.java]: you don't need to fetch those strings from the resourceBundle on init. I expect the number of feedbacks sent to be much lower than the times the form gets opened, so you should deplay the resourceBundle.getString() to when the dialog actually opens

**Comment by julianseeger** [comment]
Make sure your tests read like AAA (Arrange, Act, Assert) ... or Given, When, Then if you like
```JAVA
// arrange
setupStuff(importantVariable="significant value");
// act
String result = subjectUnderTest.executeAction();
// assert
assertEquals("significant value", result);
```
even if the method names don't make sense (like in this example), the flow and re-occurence of the tested values should make the intent and flow of the test very clear.
Act without assert is no test. And if you can delete your prouction code but the tests stay green (like `controller.key = "value"; assertEquals("value", controller.key);`, it's useless (and as far from TDD as it can be)

File: src/test-gui/java/de/qabel/desktop/ui/feedback/FeedbackPage.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    DialogPane getDialogPane() {
+        return controller.alert.getDialogPane();
+    }
+
+
+    private void typeFeedbackInformations() {
+        controller.infoMessage = "guitest infoMessage";
+        controller.titleBar = "guitest titleBar";
```
Comment (julianseeger) [inline src/test-gui/java/de/qabel/desktop/ui/feedback/FeedbackPage.java]: you don't need this (and it is not obvious for the caller)

File: src/test/java/de/qabel/desktop/ui/feedback/FeedbackControllerTest.java [hunk 9]
```diff
@@ -5,10 +5,18 @@
...
 import java.util.Locale;
 
+import static de.qabel.desktop.AsyncUtils.assertAsync;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
 
 public class FeedbackControllerTest extends AbstractControllerTest {
     FeedbackController controller;
```
Comment (julianseeger) [inline src/test/java/de/qabel/desktop/ui/feedback/FeedbackControllerTest.java]: private

**Comment by jmt85** [comment]
pls merge

File: src/main/java/de/qabel/desktop/ui/feedback/FeedbackController.java [hunk 10]
```diff
@@ -28,38 +28,50 @@
 
     @Inject
     private CrashReportHandler reportHandler;
+    private ResourceBundle resourceBundle;
+    String infoMessage;
+    String titleBar;
```
Comment (JoGir) [inline src/main/java/de/qabel/desktop/ui/feedback/FeedbackController.java]: titleBar is a bit misleading. sounds like a UI Object. may call it infoTitle or something...

File: src/main/java/de/qabel/desktop/ui/feedback/FeedbackController.java [hunk 11]
```diff
@@ -28,38 +28,50 @@
...
-
-
-
 
+        }.start();
     }
-}
 
+    void showThanksDialog() {
+        alert = new Alert(Alert.AlertType.INFORMATION, "", ButtonType.OK);
```
Comment (JoGir) [inline src/main/java/de/qabel/desktop/ui/feedback/FeedbackController.java]: Alter reference can be local

File: src/main/resources/ui.properties [hunk 12]
```diff
@@ -181,3 +181,5 @@ sourceSansLicense=Source Sans Pro
 sourceSansUrl=http://scripts.sil.org/OFL
 contactImport=Import Contacts
 contactExport=Export Contacts
+feedBackInfoMessage=Thank you for your feedback
+feedBackInfoHeader=Send feedback was successful
```
Comment (audax) [inline src/main/resources/ui.properties]: "Feedback was successfully sent"

File: src/test-gui/java/de/qabel/desktop/ui/feedback/FeedbackGuiTest.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+public class FeedbackGuiTest extends AbstractGuiTest<FeedbackController> {
+    private FeedbackPage page;
+
+    @Override
+    @Before
```
Comment (audax) [inline src/test-gui/java/de/qabel/desktop/ui/feedback/FeedbackGuiTest.java]: \@Before is not needed

**Comment by audax** [comment]
Merge with master @nnice

**Comment by julianseeger** [comment]
Mach bitte noch diesen Header "Meldung" weg. Das passt irgendwie nicht:
image

**Pull request merged**